src/sinks/util/rusoto.rs

```diff
@@ -93,12 +105,177 @@ impl Future for AwsCredentialsProviderFuture {
+        for h in request.headers().iter() {
+            let header_name = match h.0.parse::<HeaderName>() {
+                Ok(name) => name,
+                Err(err) => {
+                    return Box::new(future::err(HttpDispatchError::new(format!(
+                        "ParseHeader: {}",
+                        err
+                    ))))
+                }
+            };
+            for v in h.1.iter() {
+                let header_value = match HeaderValue::from_bytes(v) {
+                    Ok(value) => value,
+                    Err(err) => {
+                        return Box::new(future::err(HttpDispatchError::new(format!(
+                            "HeaderValueParse: {}",
+                            err
+                        ))))
+                    }
+                };
+                headers.append(&header_name, header_value);
+            }
+        }
+
+        let mut uri = format!(
+            "{}://{}{}",
+            request.scheme(),
+            request.hostname(),
+            request.canonical_path()
+        );
+
+        if !request.canonical_query_string().is_empty() {
+            uri += &format!("?{}", request.canonical_query_string());
+        }
+
+        let mut request = Request::builder()
+            .method(method)
+            .uri(uri)
+            .body(RusotoBody::from(request.payload))
+            .map_err(|e| format!("RequestBuildingError: {}", e))
+            .unwrap();
+
+        *request.headers_mut() = headers;
 
-    let ssl = SslConnector::builder(SslMethod::tls())?;
-    let https = HttpsConnector::with_connector(http, ssl)?;
+        let request = {
+            let mut client = self.client.clone();
+            client.call(request)
+        };
 
-    Ok(HttpClient::from_connector(https))
+        let fut = request
+            .and_then(|response| {
+                let status = response.status();
+                let headers = response
+                    .headers()
+                    .iter()
+                    .map(|(h, v)| {
+                        let value_string = v.to_str().unwrap().to_owned();
+                        (h.clone(), value_string)
+                    })
+                    .collect();
+                let body = response.into_body();
+                let body = BodyStream { body };
+
+                Ok(HttpResponse {
+                    status,
+                    headers,
+                    body: ByteStream::new(body),
+                })
+            })
+            .map_err(|e| HttpDispatchError::new(format!("DispatchError: {}", e)));
+
+        Box::new(fut)
+    }
+}
+
+impl Payload for RusotoBody {
+    type Data = io::Cursor<Vec<u8>>;
+    type Error = io::Error;
+
+    fn poll_data(&mut self) -> Poll<Option<Self::Data>, Self::Error> {
+        match &mut self.inner {
+            Some(SignedRequestPayload::Buffer(buf)) => {
+                if !buf.is_empty() {
+                    let buf = buf.split_off(0);
```

Comment:
Am I understand correctly that `split_off(0)` called here for removing data from `SignedRequestPayload::Buffer(Bytes)`? i.e. memory freeing asap?

Reply:
It extracts the contents of the buffer into the binding `buf` and replaces it with empty. That `buf` is later turned into a `Cursor` and returned.

Reply:
Yeah I understand what's happen next with `buf` content, I just curious why `split_off` here. Turning to `Cursor` can be done without `split_off` and I thought that it's here for some specific reason.

Reply:
This can't just use grab buffer directly, as we don't own it here, we only have a `mut ref`. If it were an option, the equivalent operation would be `take`. At least, that's what I'm seeing.